examples/checkout.c

```diff
@@ -0,0 +1,229 @@
+	int err;
+
+	/** Setup our checkout options from the parsed options */
+	checkout_opts.checkout_strategy = GIT_CHECKOUT_SAFE;
+	if (opts->force)
+		checkout_opts.checkout_strategy = GIT_CHECKOUT_FORCE;
+
+	if (opts->progress)
+		checkout_opts.progress_cb = print_checkout_progress;
+
+	if (opts->perf)
+		checkout_opts.perfdata_cb = print_perf_data;
+
+	/** Grab the commit we're interested to move to */
+	err = git_commit_lookup(&target_commit, repo, git_annotated_commit_id(target));
+	if (err != 0) {
+		fprintf(stderr, "failed to lookup commit: %s\n", giterr_last()->message);
+		goto cleanup;
+	}
+
+	/**
+	 * Perform the checkout so the workdir corresponds to what target_commit
+	 * contains.
+	 *
+	 * Note that it's okay to pass a git_commit here, because it will be 
+	 * peeled to a tree.
+	 */
+	err = git_checkout_tree(repo, (const git_object *)target_commit, &checkout_opts);
+	if (err != 0) {
+		fprintf(stderr, "failed to checkout tree: %s\n", giterr_last()->message);
+		goto cleanup;
+	}
+
+	/**
+	 * Now that the checkout has completed, we have to update HEAD.
+	 *
+	 * Depending on the "origin" of target (ie. it's an OID or a branch name),
+	 * we might need to detach HEAD.
+	 */
+	if (git_annotated_commit_ref(target)) {
+		err = git_repository_set_head(repo, git_annotated_commit_ref(target));
+	} else {
+		err = git_repository_set_head_detached(repo, git_annotated_commit_id(target));
+	}
+	if (err != 0) {
+		fprintf(stderr, "failed to update HEAD reference: %s\n", giterr_last()->message);
+		goto cleanup;
+	}
+
+cleanup:
+	git_commit_free(target_commit);
+
+	return err;
+}
+
+/** That example's entry point */
+int main(int argc, char **argv)
+{
+	git_repository *repo = NULL;
+	struct args_info args = ARGS_INFO_INIT;
+	checkout_options opts;
+	git_repository_state_t state;
+	git_annotated_commit *checkout_target = NULL;
+	int err = 0;
+	const char *path = ".";
+
+	/** Parse our command line options */
+	parse_options(&path, &opts, &args);
+
+	/** Initialize the library */
+	err = git_libgit2_init();
+	if (!err)
+		check_lg2(err, "Failed to initialize libgit2", NULL);
+
+	/** Open the repository corresponding to the options */
+	check_lg2(git_repository_open_ext(&repo, path, 0, NULL),
+			  "Could not open repository", NULL);
+
+	/** Make sure we're not about to checkout while something else is going on */
+	state = git_repository_state(repo);
+	if (state != GIT_REPOSITORY_STATE_NONE) {
+		fprintf(stderr, "repository is in unexpected state %d\n", state);
+		goto cleanup;
+	}
+
+	if (args.pos >= args.argc) {
+		fprintf(stderr, "unhandled\n");
+		err = -1;
+		goto cleanup;
+	} else if (strcmp("--", args.argv[args.pos])) {
+		/**
+		 * Try to checkout the given path
+		 */
+
+		fprintf(stderr, "unhandled path-based checkout\n");
+		err = 1;
+		goto cleanup;
+	} else {
+		/**
+		 * Try to resolve a "refish" argument to a target libgit2 can use
+		 */
+		err = resolve_refish(&checkout_target, repo, args.argv[args.pos]);
```

Comment:
Resolving the ref here makes sense, but I'd parse the reference itself into the  options structure in `parse_options`